### PR TITLE
Add support for the `--include` flag

### LINKED ISSUE #219: Idea: --only flag?
_I have just discovered Madge and it's just superb work!_

It would be interesting for me, as I work on a Git branch, to understand how the files I've created relate to each other:

```bash
git diff --name-only --diff-filter=A origin/master...HEAD | grep -E '\.js$' | xargs madge
```

This command takes the added files as entry points, but shows their entire dependency graph.  It would be great to limit it to the files I've sent as entry points. 

If I understand correctly, the `exclude` option accepts a RegExp and I don't think I can use it.

Would an `--only` option make sense? It would limit the set of consider files to the original input. Or, if that's too specific, maybe an `--include` option (that you can use multiple time in a command), to supplement `--exclude`?

### PR DESCRIPTION
**Pull request by danburzo**
Fixes #219. This is a redo of a previous PR:

- https://github.com/pahen/madge/pull/220

### CODE REVIEW

**Comment by danburzo** [comment]
No worries, I'll be glad to include some tests when I find a window of time! 

In regards to [your proposal](https://github.com/pahen/madge/pull/220#issuecomment-1408136437) to mirror `.tsconfig` conventions, I think because `--exclude` already works differently from `.tsconfig.exclude`, making an analogy is not possible. 

> Important: exclude only changes which files are [include](https://www.typescriptlang.org/tsconfig#include)d as a result of the include setting. A file specified by exclude can still become part of your codebase due to an import statement in your code, a types inclusion, a /// <reference directive, or being specified in the [files](https://www.typescriptlang.org/tsconfig#files) list.
> 
> It is not a mechanism that prevents a file from being [include](https://www.typescriptlang.org/tsconfig#include)d in the codebase - it simply changes what the include setting finds.

I'll rethink the names of the CLI options to see if there's something better available.

**Comment by PabloLION** [comment]
I agree. The current `dependencyFilter` is kinda messy (see #282) so it's probable that we change everything to mirror `tsconfig`.
I'm not sure how current `--exclude` differs from `tsconfig`. For me they both exclude some "modules" to be parsed/checked by tsc/madge (because `tsconfig` has `include` default to "[] if [files](https://www.typescriptlang.org/tsconfig#files) is specified, `**` otherwise." ) 

One difference I noticed is the glob-pattern and RegExp. I can't decide which one is better:
- RegExp is much more expressive
- glob-pattern has less version / flavor, hence more stable

Maybe we can first keep everything the same to finish this PR? We welcome all to discuss about the new `--include` and `--exclude` in a new issue.

**Comment by danburzo** [comment]
> I'm not sure how current `--exclude` differs from `tsconfig`

If I understand correctly (not a TS user), `.tsconfig`'s `include`, `exclude`, and `files` options control the set of input files. All `import`s are then added to the list of processing files. 

In contrast, `include` and `exclude` as defined in this PR (as well as the current definition of `exclude`) act together as a filter to what's rendered in the graph. The set of input files are defined as the operands to the CLI interface: `madge file1 file2 ...` and expanded based on their dependencies, then filtered by `include`/`exclude`: they have to match an `include` pattern and not be excluded by an `exclude` pattern.

File: lib/tree.js [hunk 1]
```diff
@@ -236,15 +236,18 @@ class Tree {
...
 	 */
-	exclude(tree, excludeRegExp) {
-		const regExpList = excludeRegExp.map((re) => new RegExp(re));
+	exclude(tree, includeRegExp, excludeRegExp) {
+		const includeList = includeRegExp ? includeRegExp.map((re) => new RegExp(re)) : false;
+		const excludeList = excludeRegExp ? excludeRegExp.map((re) => new RegExp(re)) : false;
 
 		function regExpFilter(id) {
-			return regExpList.findIndex((regexp) => regexp.test(id)) < 0;
+			return (!includeList || includeList.findIndex((re) => re.test(id)) >= 0) &&
```
Comment (PabloLION) [inline lib/tree.js:249]: This logic means, only filenames declared in `--include` are included, other files, even those not `--exclude`-d, won't be checked?

**Comment by PabloLION** [comment]
@danburzo, sorry 😅 I was too stupid. I just realized: in `tsconfig`, `include` and `exclude` only changes the entry points rather than files transpiled by typescript.

What I wanted to say at first was to give priorities to `include` and `exclude`: like `tsconfig.json` first add all files whose filename fits the `include` pattern, then remove those of `exclude`.
Maybe we should implement something like this?

**Pull request closed without merging**